### PR TITLE
Overhaul monkeypatching again

### PR DESCRIPTION
**Pull request by rbtcollins**
This change is [Reviewable](https://reviewable.io/reviews/testing-cabal/fixtures/30)

### CODE REVIEW

File: fixtures/tests/_fixtures/test_monkeypatch.py [hunk 1]
```diff
@@ -310,16 +310,17 @@ def test_patch_boundmethod_with_boundmethod(self):
...
     def test_patch_unboundmethod_with_staticmethod(self):
         oldmethod = C.foo
         fixture = MonkeyPatch(
             'fixtures.tests._fixtures.test_monkeypatch.C.foo',
             D.bar_static_args)
         with fixture:
-            c = C() 
-            tgtslf, arg = c.foo(1)
-            self.expectThat(tgtslf, Is(c))
+            tgtslf, arg = INST_C.foo(1)
```
Comment (alaski) [inline fixtures/tests/_fixtures/test_monkeypatch.py]: Was this change intentional? The class is being patched and the test checks against an instance of that class which was instantiated before patching. But the way this is done provides no check that INST_C.foo is actually patched and not the original method.

**Comment by alaski** [comment]
This is looking really good. I appreciate the cleanups. Just one comment inline.

File: fixtures/tests/_fixtures/test_monkeypatch.py (hunk 1 above)
Comment (rbtcollins) [inline fixtures/tests/_fixtures/test_monkeypatch.py]: Not sure what you mean; the test is patching C.foo - the unbound method, and the original C.foo would return arg, not (self, arg), so we know it is indeed the patched in method (or a failed unpatch from an earlier test, I guess). Happy to change it to make a C() here again - this was how I discovered the failed cleanups of patching bound methods.

Comment (alaski) [inline fixtures/tests/_fixtures/test_monkeypatch.py]: No, you're right. I missed the different return.

**Comment by alaski** [comment]
Okay, I'm satisfied with this at the moment.

File: fixtures/_fixtures/monkeypatch.py [hunk 2]
```diff
@@ -17,27 +17,70 @@
...
+    _class_types = _class_types + (types.ClassType,)
 
-    setattr transforms a function into instancemethod, so where appropriate
-    value needs to be wrapped with staticmethod().
+
+def _coerce_values(obj, name, new_value, sentinel):
+    """Handle value coercion for static and classmethods.
+
+    setattr transforms a function into an instancemethod when set on a class.
+    This checks if the attribute to be replaced is either and wraps new_value
```
Comment (thomir) [inline fixtures/_fixtures/monkeypatch.py]: is either what? The previous sentence contains three possible nouns (function, instancemethod, class). This doesn't make any sense to me. Can you re-word please?

File: fixtures/_fixtures/monkeypatch.py [hunk 3]
```diff
@@ -17,27 +17,70 @@
...
+    if necessary. This also checks getattr(obj, name) and wraps it if necessary
+    since the staticmethod wrapper isn't preserved.
```
Comment (thomir) [inline fixtures/_fixtures/monkeypatch.py]: Might also be worth documenting the return values from this function? It seems it returns the (new_value, old_value) unless the thing we're patching doesn't exist already, in which case we return (new_value, sentinel)?

File: fixtures/tests/_fixtures/test_monkeypatch.py [hunk 4]
```diff
@@ -72,14 +108,318 @@ def test_delete_missing_attribute(self):
...
         fixture = MonkeyPatch(
             'fixtures.tests._fixtures.test_monkeypatch.C.foo',
-            bar)
+            D.bar_static_args)
         with fixture:
-            C.foo()
-            C().foo()
-        self.assertEqual(oldfoo, C.foo)
-        self.assertEqual(oldfoo, C().foo)
+            tgtslf, arg = INST_C.foo(1)
```
Comment (thomir) [inline fixtures/tests/_fixtures/test_monkeypatch.py]: tgtslf? Can we make this a little more explicit and meaningful please?

File: fixtures/tests/_fixtures/test_monkeypatch.py [hunk 5]
```diff
@@ -72,14 +108,318 @@ def test_delete_missing_attribute(self):
...
+        oldmethod_inst = C().foo
+        fixture = MonkeyPatch(
+            'fixtures.tests._fixtures.test_monkeypatch.C.foo',
+            D().bar_self_referential)
+        with fixture:
+            C().foo()
+        self.assertEqual(oldmethod, C.foo)
+        # The method address changes with each instantiation of C, and method
+        # equivalence just tests that. Compare the code objects instead.
+        self.assertEqual(oldmethod_inst.__code__, C().foo.__code__)
```
Comment (thomir) [inline fixtures/tests/_fixtures/test_monkeypatch.py]: Given that this assertion happens at least twice, is it worth making a custom Matcher or assertion method? That way you could probably get rid of the comment as well.

**Comment by thomir** [comment]
This looks good, but there are a few please where I think the intent can be made clearer - see diff comments. I'm +1 regardless however.
